codegen/src/main/scala/bsp/codegen/TypescriptSpecRenderer.scala

```diff
@@ -0,0 +1,121 @@
+package bsp.codegen
+
+import bsp.codegen.Def._
+import bsp.codegen.Primitive._
+import bsp.codegen.Type._
+import cats.syntax.all._
+import software.amazon.smithy.model.shapes.ShapeId
+
+import dsl._
+import bsp.codegen.EnumType.IntEnum
+import bsp.codegen.EnumType.StringEnum
+import bsp.codegen.JsonRPCMethodType.Notification
+import bsp.codegen.JsonRPCMethodType.Request
+import bsp.codegen.Hint.Documentation
+
+class TypescriptSpecRenderer(basepkg: String) {
```

Comment:
Rendering TS may come in handy anyway, but I wonder if we should just migrate the main spec to Smithy, so there's less of a disconnect between source of truth and documentation. Is there a straightforward way of doing that?

Reply:
You mean advertise the smithy file as documentation ? Or generate markdown with smithy snippets in it ?

Either way, I think the best course of action is to defer that decision. This POC should, ihmo, strive to recreate something reasonably close to the current state. 

Smithy is not as well known as typescript by any mean, and offering documentation with typescript snippets kinda make sense, in particular if those snippets are automatically rendered as opposed to handcrafted.

If later on you want to make it more dynamic and render the snippets in the markdown differently, it'd be possible. I think this PR proves that code generation from smithy is not difficult, whatever the target language is.

Reply:
Smithy provides a renderer out of the box but the atomic unit of rendering you get is the namespace : the function that renders individual shape is private. You could ask Aws whether they'd be keen to make it public but I have no idea whether they'd agree to it.

Regarding this renderer, I'm writing it to use for markdown but you could obviously change the code in slight ways to render an actual typescript library